File: scripts/rename-api-paths.py
#!/usr/bin/env python3
"""Rename list/delete API paths to {biz}PageList and remove{Biz} per controller."""
from __future__ import annotations

import re
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]

BASE_ENTITY: dict[str, str] = {
    "/sys/user": "user",
    "/sys/role": "role",
    "/sys/post": "post",
    "/sys/dept": "dept",
    "/sys/menu": "menu",
    "/sys/operlog": "operLog",
    "/sys/logininfor": "logininfor",
    "/sys/dict/type": "dictType",
    "/sys/dict/data": "dictData",
    "/config": "config",
    "/notice": "notice",
    "/avatar": "avatar",
    "/mall/product": "product",
    "/mall/sku": "sku",
    "/mall/category": "category",
    "/mall/attr": "attr",
    "/mall/group": "attrGroup",
    "/mall/brand": "brand",
    "/mall/comment": "comment",
    "/mall/coupon": "coupon",
    "/mall/coupon-history": "couponHistory",
    "/mall/member-price": "memberPrice",
    "/mall/spu-bounds": "spuBounds",
    "/mall/sku-ladder": "skuLadder",
    "/mall/sku-full-reduction": "skuFullReduction",
    "/mall/home-adv": "homeAdv",
    "/mall/subject": "subject",
    "/mall/seckill-session": "seckillSession",
    "/mall/seckill-promotion": "seckillPromotion",
    "/mall/member": "member",
    "/mall/member-level": "memberLevel",
    "/mall/member-statistics": "memberStatistics",
    "/mall/address": "address",
    "/mall/wareinfo": "wareInfo",
    "/mall/ware-sku": "wareSku",
    "/mall/ware-task": "wareTask",
    "/mall/purchase": "purchase",
    "/mall/order": "order",
    "/mall/order-return": "orderReturn",
    "/mall/payment": "payment",
    "/mall/refund": "refund",
    "/portal/order": "portalOrder",
    "/portal/cart": "cart",
    "/approval/template": "template",
    "/approval/instance": "approvalInstance",
    "/approval/task": "approvalTask",
    "/file": "file",
    "/tool/gen": "genTable",
    "/monitor/job": "job",
}

SUBPATH_RENAMES: dict[str, list[tuple[str, str]]] = {
    "/mall/member-growth": [
        ("/integration/list", "/integrationPageList"),
        ("/growth/list", "/growthPageList"),
    ],
    "/approval/instance": [
        ("/mine/list", "/mineInstancePageList"),
    ],
    "/approval/task": [
        ("/todo/list", "/todoTaskPageList"),
        ("/done/list", "/doneTaskPageList"),
    ],
    "/approval/template": [
        ("/bind/list", "/templateBindPageList"),
        ("/bind/remove", "/removeTemplateBind"),
    ],
    "/mall/purchase": [
        ("/unreceive/list", "/unreceivePurchasePageList"),
        ("/detail/list", "/purchaseDetailPageList"),
        ("/detail/remove", "/removePurchaseDetail"),
    ],
    "/tool/gen": [
        ("/db/list", "/genDbTablePageList"),
    ],
    "/monitor/job": [
        ("/log/list", "/jobLogPageList"),
    ],
    "/file": [
        ("/recycle/list", "/recycleFilePageList"),
    ],
}

REQUEST_MAPPING_RE = re.compile(r'@RequestMapping\("([^"]+)"\)')
POST_MAPPING_RE = re.compile(r'@PostMapping\("([^"]+)"\)')
DELETE_MAPPING_RE = re.compile(r'@DeleteMapping\("([^"]+)"\)')
POST_MAPPING_MULTI_RE = re.compile(r'@PostMapping\(\{([^}]+)\}\)')
DELETE_MAPPING_MULTI_RE = re.compile(r'@DeleteMapping\(\{([^}]+)\}\)')


def pascal(camel: str) -> str:
    return camel[0].upper() + camel[1:] if camel else camel


def infer_entity(base: str) -> str:
    if base in BASE_ENTITY:
        return BASE_ENTITY[base]
    seg = base.rstrip("/").split("/")[-1]
    if "-" in seg:
        parts = seg.split("-")
        return parts[0] + "".join(p[:1].upper() + p[1:] for p in parts[1:])
    return seg


def url_replacements() -> list[tuple[str, str]]:
    reps: list[tuple[str, str]] = []
    for base, entity in {**BASE_ENTITY, **{b: infer_entity(b) for b in SUBPATH_RENAMES}}.items():
        if base not in BASE_ENTITY and base not in SUBPATH_RENAMES:
            continue
    for base, entity in BASE_ENTITY.items():
        page = f"/{entity}PageList"
        remove = f"/remove{pascal(entity)}"
        for old in ("/list", "/pageList"):
            reps.append((f"{base}{old}", f"{base}{page}"))
        for old in ("/delete", "/remove", "/removeNotice"):
            reps.append((f"{base}{old}", f"{base}{remove}"))
    for base, mappings in SUBPATH_RENAMES.items():
        for old, new in mappings:
            reps.append((f"{base}{old}", f"{base}{new}"))
    reps.append(("/mall/member-level/list", "/mall/member-level/memberLevelPageList"))
    return sorted(set(reps), key=lambda x: len(x[0]), reverse=True)


def transform_controller(text: str) -> tuple[str, list[tuple[str, str]]]:
    m = REQUEST_MAPPING_RE.search(text)
    if not m:
        return text, []
    base = m.group(1)
    entity = infer_entity(base)
    page = f"/{entity}PageList"
    remove = f"/remove{pascal(entity)}"
    local_reps: list[tuple[str, str]] = []

    def sub_post(match: re.Match[str]) -> str:
        inner = match.group(1)
        if "list" in inner or "pageList" in inner:
            local_reps.append((f"{base}/list", f"{base}{page}"))
            local_reps.append((f"{base}/pageList", f"{base}{page}"))
            return f'@PostMapping("{page}")'
        return match.group(0)

    def sub_delete(match: re.Match[str]) -> str:
        inner = match.group(1)
        if "delete" in inner or "remove" in inner:
            local_reps.append((f"{base}/delete", f"{base}{remove}"))
            local_reps.append((f"{base}/remove", f"{base}{remove}"))
            local_reps.append((f"{base}/removeNotice", f"{base}{remove}"))
            return f'@DeleteMapping("{remove}")'
        return match.group(0)

    text = POST_MAPPING_MULTI_RE.sub(sub_post, text)
    text = DELETE_MAPPING_MULTI_RE.sub(sub_delete, text)

    for old, new in [
        ('@PostMapping("/pageList")', f'@PostMapping("{page}")'),
        ('@PostMapping("/list")', f'@PostMapping("{page}")'),
        ('@DeleteMapping("/delete")', f'@DeleteMapping("{remove}")'),
        ('@DeleteMapping("/remove")', f'@DeleteMapping("{remove}")'),
        ('@DeleteMapping("/removeNotice")', f'@DeleteMapping("{remove}")'),
    ]:
        if old in text:
            if 'PostMapping' in old:
                suffix_old = "/pageList" if "pageList" in old else "/list"
                local_reps.append((f"{base}{suffix_old}", f"{base}{page}"))
            else:
                for s in ("/delete", "/remove", "/removeNotice"):
                    if s in old:
                        local_reps.append((f"{base}{s}", f"{base}{remove}"))
            text = text.replace(old, new)

    for old_suffix, new_suffix in SUBPATH_RENAMES.get(base, []):
        post_old = f'@PostMapping("{old_suffix}")'
        post_new = f'@PostMapping("{new_suffix}")'
        del_old = f'@DeleteMapping("{old_suffix}")'
        del_new = f'@DeleteMapping("{new_suffix}")'
        if post_old in text:
            local_reps.append((f"{base}{old_suffix}", f"{base}{new_suffix}"))
            text = text.replace(post_old, post_new)
        if del_old in text:
            local_reps.append((f"{base}{old_suffix}", f"{base}{new_suffix}"))
            text = text.replace(del_old, del_new)

    # member-level: GET /list -> keep; POST /list already handled
    if base == "/mall/member-level":
        local_reps.append(("/mall/member-level/list", "/mall/member-level/memberLevelPageList"))

    return text, local_reps


def main() -> None:
    url_reps = url_replacements()
    java_roots = [
        ROOT / "starpivot-system",
        ROOT / "starpivot-mall",
        ROOT / "starpivot-approval",
        ROOT / "starpivot-file",
        ROOT / "starpivot-generator",
        ROOT / "starpivot-monitor",
    ]
    changed_java = 0
    for root in java_roots:
        for path in root.rglob("*Controller.java"):
            text = path.read_text(encoding="utf-8")
            new_text, _ = transform_controller(text)
            if new_text != text:
                path.write_text(new_text, encoding="utf-8")
                changed_java += 1
                print(f"java: {path.relative_to(ROOT)}")

    ui_api = ROOT / "star-pivot-ui" / "src" / "api"
    changed_ts = 0
    if ui_api.exists():
        for path in ui_api.rglob("*.ts"):
            text = path.read_text(encoding="utf-8")
            original = text
            for old, new in url_reps:
                text = text.replace(f"/api{old}", f"/api{new}")
            if text != original:
                path.write_text(text, encoding="utf-8")
                changed_ts += 1
                print(f"ts: {path.relative_to(ROOT)}")

    print(f"Done. Java: {changed_java}, TS: {changed_ts}")


if __name__ == "__main__":
    main()

File: scripts/rename-api-paths.mjs
#!/usr/bin/env node
/** Rename list/delete API paths to {biz}PageList and remove{Biz} per controller. */
import { readFileSync, writeFileSync, readdirSync, statSync } from 'fs';
import { join, relative } from 'path';
import { fileURLToPath } from 'url';

const __dirname = fileURLToPath(new URL('.', import.meta.url));
const ROOT = join(__dirname, '..');

const BASE_ENTITY = {
  '/sys/user': 'user',
  '/sys/role': 'role',
  '/sys/post': 'post',
  '/sys/dept': 'dept',
  '/sys/menu': 'menu',
  '/sys/operlog': 'operLog',
  '/sys/logininfor': 'logininfor',
  '/sys/dict/type': 'dictType',
  '/sys/dict/data': 'dictData',
  '/config': 'config',
  '/notice': 'notice',
  '/avatar': 'avatar',
  '/mall/product': 'product',
  '/mall/sku': 'sku',
  '/mall/category': 'category',
  '/mall/attr': 'attr',
  '/mall/group': 'attrGroup',
  '/mall/brand': 'brand',
  '/mall/comment': 'comment',
  '/mall/coupon': 'coupon',
  '/mall/coupon-history': 'couponHistory',
  '/mall/member-price': 'memberPrice',
  '/mall/spu-bounds': 'spuBounds',
  '/mall/sku-ladder': 'skuLadder',
  '/mall/sku-full-reduction': 'skuFullReduction',
  '/mall/home-adv': 'homeAdv',
  '/mall/subject': 'subject',
  '/mall/seckill-session': 'seckillSession',
  '/mall/seckill-promotion': 'seckillPromotion',
  '/mall/member': 'member',
  '/mall/member-level': 'memberLevel',
  '/mall/member-statistics': 'memberStatistics',
  '/mall/address': 'address',
  '/mall/wareinfo': 'wareInfo',
  '/mall/ware-sku': 'wareSku',
  '/mall/ware-task': 'wareTask',
  '/mall/purchase': 'purchase',
  '/mall/order': 'order',
  '/mall/order-return': 'orderReturn',
  '/mall/payment': 'payment',
  '/mall/refund': 'refund',
  '/portal/order': 'portalOrder',
  '/portal/cart': 'cart',
  '/approval/template': 'template',
  '/approval/instance': 'approvalInstance',
  '/approval/task': 'approvalTask',
  '/file': 'file',
  '/tool/gen': 'genTable',
  '/monitor/job': 'job',
  '/monitor/cache': 'cache',
};

const SUBPATH_RENAMES = {
  '/mall/member-growth': [
    ['/integration/list', '/integrationPageList'],
    ['/growth/list', '/growthPageList'],
  ],
  '/approval/instance': [['/mine/list', '/mineInstancePageList']],
  '/approval/task': [
    ['/todo/list', '/todoTaskPageList'],
    ['/done/list', '/doneTaskPageList'],
  ],
  '/approval/template': [
    ['/bind/list', '/templateBindPageList'],
    ['/bind/remove', '/removeTemplateBind'],
  ],
  '/mall/purchase': [
    ['/unreceive/list', '/unreceivePurchasePageList'],
    ['/detail/list', '/purchaseDetailPageList'],
    ['/detail/remove', '/removePurchaseDetail'],
  ],
  '/tool/gen': [['/db/list', '/genDbTablePageList']],
  '/monitor/job': [['/log/list', '/jobLogPageList']],
  '/file': [['/recycle/list', '/recycleFilePageList']],
};

const REQUEST_MAPPING_RE = /@RequestMapping\("([^"]+)"\)/;
const POST_MAPPING_MULTI_RE = /@PostMapping\(\{([^}]+)\}\)/g;
const DELETE_MAPPING_MULTI_RE = /@DeleteMapping\(\{([^}]+)\}\)/g;

function pascal(camel) {
  return camel ? camel[0].toUpperCase() + camel.slice(1) : camel;
}

function inferEntity(base) {
  if (BASE_ENTITY[base]) return BASE_ENTITY[base];
  const seg = base.replace(/\/$/, '').split('/').pop();
  if (seg.includes('-')) {
    const parts = seg.split('-');
    return parts[0] + parts.slice(1).map((p) => p[0].toUpperCase() + p.slice(1)).join('');
  }
  return seg;
}

function urlReplacements() {
  const reps = [];
  for (const [base, entity] of Object.entries(BASE_ENTITY)) {
    const page = `/${entity}PageList`;
    const remove = `/remove${pascal(entity)}`;
    for (const old of ['/list', '/pageList']) {
      reps.push([`${base}${old}`, `${base}${page}`]);
    }
    for (const old of ['/delete', '/remove', '/removeNotice']) {
      reps.push([`${base}${old}`, `${base}${remove}`]);
    }
  }
  for (const [base, mappings] of Object.entries(SUBPATH_RENAMES)) {
    for (const [old, neu] of mappings) {
      reps.push([`${base}${old}`, `${base}${neu}`]);
    }
  }
  reps.push(['/mall/member-level/list', '/mall/member-level/memberLevelPageList']);
  return [...new Map(reps.map((r) => [r[0], r[1]])).entries()].sort((a, b) => b[0].length - a[0].length);
}

function transformController(text) {
  const m = text.match(REQUEST_MAPPING_RE);
  if (!m) return text;
  const base = m[1];
  const entity = inferEntity(base);
  const page = `/${entity}PageList`;
  const remove = `/remove${pascal(entity)}`;

  text = text.replace(POST_MAPPING_MULTI_RE, (match, inner) => {
    if (inner.includes('list') || inner.includes('pageList')) {
      return `@PostMapping("${page}")`;
    }
    return match;
  });

  text = text.replace(DELETE_MAPPING_MULTI_RE, (match, inner) => {
    if (inner.includes('delete') || inner.includes('remove')) {
      return `@DeleteMapping("${remove}")`;
    }
    return match;
  });

  const singles = [
    ['@PostMapping("/pageList")', `@PostMapping("${page}")`],
    ['@PostMapping("/list")', `@PostMapping("${page}")`],
    ['@GetMapping("/list")', `@GetMapping("${page}")`],
    ['@DeleteMapping("/delete")', `@DeleteMapping("${remove}")`],
    ['@DeleteMapping("/remove")', `@DeleteMapping("${remove}")`],
    ['@DeleteMapping("/removeNotice")', `@DeleteMapping("${remove}")`],
  ];
  for (const [old, neu] of singles) {
    text = text.split(old).join(neu);
  }

  for (const [oldSuffix, newSuffix] of SUBPATH_RENAMES[base] ?? []) {
    text = text.split(`@PostMapping("${oldSuffix}")`).join(`@PostMapping("${newSuffix}")`);
    text = text.split(`@DeleteMapping("${oldSuffix}")`).join(`@DeleteMapping("${newSuffix}")`);
  }

  return text;
}

function walk(dir, suffix, out = []) {
  for (const name of readdirSync(dir)) {
    const p = join(dir, name);
    if (statSync(p).isDirectory()) walk(p, suffix, out);
    else if (name.endsWith(suffix)) out.push(p);
  }
  return out;
}

const urlReps = urlReplacements();
const javaRoots = [
  'starpivot-system',
  'starpivot-mall',
  'starpivot-approval',
  'starpivot-file',
  'starpivot-generator',
  'starpivot-monitor',
];

let changedJava = 0;
for (const root of javaRoots) {
  const abs = join(ROOT, root);
  for (const path of walk(abs, 'Controller.java')) {
    const text = readFileSync(path, 'utf8');
    const newText = transformController(text);
    if (newText !== text) {
      writeFileSync(path, newText, 'utf8');
      changedJava++;
      console.log(`java: ${relative(ROOT, path)}`);
    }
  }
}

const uiApi = join(ROOT, 'star-pivot-ui', 'src', 'api');
let changedTs = 0;
if (statSync(uiApi, { throwIfNoEntry: false })) {
  for (const path of walk(uiApi, '.ts')) {
    let text = readFileSync(path, 'utf8');
    const original = text;
    for (const [old, neu] of urlReps) {
      const escaped = old.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
      text = text.replace(new RegExp(`/api${escaped}(?=["'\`])`, 'g'), `/api${neu}`);
    }
    if (text !== original) {
      writeFileSync(path, text, 'utf8');
      changedTs++;
      console.log(`ts: ${relative(ROOT, path)}`);
    }
  }
}

console.log(`Done. Java: ${changedJava}, TS: ${changedTs}`);
